lldb/source/Target/ThreadPlanStepThrough.cpp

```diff
@@ -90,8 +90,10 @@ void ThreadPlanStepThrough::LookForPlanToStepThroughFromCurrentPC() {
       m_sub_plan_sp =
           runtime->GetStepThroughTrampolinePlan(thread, m_stop_others);
 
-      if (m_sub_plan_sp)
+      if (m_sub_plan_sp && m_sub_plan_sp->ValidatePlan(nullptr))
         break;
+
+      m_sub_plan_sp.reset();
```

Comment:
@jimingham should this be upstreamed to llvm? On one hand, it's not needed there, on the other hand I think it's good since in this codepath `ValidatePlan` isn't checked. Or maybe the `ValidatePlan` should happen somewhere else?

Reply:
QueueThreadPlan, which is what is called by most external clients, does call ValidatePlan.  So that's how it gets done from the outside.  But in this case the API is returning a plan it found to do a step through, so it makes more .sense to validate the plan before handing it out.  Otherwise you are saying "I found a step through plan" when what you actually found was a non-working step-through plan.  That doesn't seem very helpful.

Reply:
The reason I added this check is because I wanted to encapsulate the relevant logic within `ThreadPlanStepInAsync`. When no more stepping is needed, I put that decision should be within the plan rather than outside the plan. I saw it as a no-op plan rather than a non-working plan, but I see now that returning false from `ValidatePlan` can be saying it's non-working. I'll refactor it.

Reply:
Restored this code. I added a bool function on `ThreadPlanStepInAsync` that does a preflight check, to avoid a no-op/non-working plan.